src/main/java/com/google/sps/servlets/RecipeServlet.java

```diff
@@ -73,121 +81,129 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
       return;
     }
 
     String token = request.getParameter("token");
-    FirebaseToken decodedToken = Auth.verifyIdToken(token);
-    if (decodedToken == null) {
-      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
-      return;
-    }
-    if (!User.createdRecipe(decodedToken.getUid(), newRecipe.id)) {
-      response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+    String uid = matchUser(token, newRecipe.metadata.id, response);
+    if (uid == null) {
       return;
     }
 
-    DocumentReference recipeRef = DBUtils.recipe(newRecipe.id);
-    ApiFuture future =
-        recipeRef.update(Recipe.TITLE_KEY, newRecipe.title, Recipe.CONTENT_KEY, newRecipe.content);
-    DBUtils.blockOnFuture(future);
+    dbInterface.editRecipeTitleContent(
+        newRecipe.metadata.id, newRecipe.metadata.title, newRecipe.content);
   }
 
   @Override
   public void doDelete(HttpServletRequest request, HttpServletResponse response)
       throws IOException {
-    String recipeID = request.getParameter("recipeID");
-    if (recipeID == null) {
+    String recipeId = request.getParameter("recipeID");
+    if (recipeId == null) {
       response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
       return;
     }
 
     String token = request.getParameter("token");
-    FirebaseToken decodedToken = Auth.verifyIdToken(token);
-    if (decodedToken == null) {
-      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
-      return;
-    }
-    if (!User.createdRecipe(decodedToken.getUid(), recipeID)) {
-      response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+    String uid = matchUser(token, recipeId, response);
+    if (uid == null) {
       return;
     }
 
-    deleteComments(recipeID);
-    ApiFuture<WriteResult> writeResult = DBUtils.recipes().document(recipeID).delete();
-    DBUtils.blockOnFuture(writeResult);
+    dbInterface.deleteComments(recipeId);
+    dbInterface.deleteRecipe(recipeId);
   }
 
-  private String getRecipeList() {
-    Query query = DBUtils.recipes();
-    ApiFuture<QuerySnapshot> querySnapshotFuture = query.get();
-    ArrayList<Object> recipeList = new ArrayList<>();
-    QuerySnapshot querySnapshot = DBUtils.blockOnFuture(querySnapshotFuture);
-
-    if (querySnapshot == null) {
-      return null;
-    }
-
-    for (DocumentSnapshot document : querySnapshot.getDocuments()) {
-      recipeList.add(document.getData());
+  protected String getRecipeList(HttpServletRequest request, HttpServletResponse response) {
+    String creatorToken = request.getParameter("token");
+
+    String tagParam = request.getParameter("tagIDs");
+    boolean isSavedRequest = Boolean.parseBoolean(request.getParameter("saved"));
+
+    boolean isTagQuery = !(tagParam == null || tagParam.equals("None"));
+    boolean isCreatorQuery = !(creatorToken == null || creatorToken.equals("None"));
+
+    if (isSavedRequest || (isCreatorQuery && !isTagQuery)) {
```

Comment:
This whole if chain is hard to decipher. What exactly is the expected behavior here? Adding a comment to explain the expected behavior is a good.

Reply:
Good point